compiler/rustc_mir/src/transform/flatten_locals.rs

```diff
@@ -0,0 +1,238 @@
+use crate::transform::MirPass;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_index::vec::Idx;
+use rustc_middle::mir::interpret::Scalar;
+use rustc_middle::mir::visit::*;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use std::collections::hash_map::Entry;
+
+pub struct FlattenLocals;
+
+impl<'tcx> MirPass<'tcx> for FlattenLocals {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        if tcx.sess.mir_opt_level() < 4 {
+            return;
+        }
+
+        let replacements = compute_flattening(tcx, body);
+        let mut all_dead_locals = FxHashSet::default();
+        all_dead_locals.extend(replacements.discr.keys().copied());
+        all_dead_locals.extend(replacements.fields.keys().map(|p| p.local));
+        if all_dead_locals.is_empty() {
+            return;
+        }
+
+        let mut visitor = FlattenVisitor { tcx, map: &replacements, all_dead_locals };
+        visitor.visit_body(body);
+        super::simplify::simplify_locals(body, tcx);
+    }
+}
+
+fn escaping_locals(body: &Body<'_>) -> FxHashSet<Local> {
+    let set = (0..body.arg_count + 1).map(Local::new).collect();
+    let mut visitor = EscapeVisitor { escaping: false, set };
+    visitor.visit_body(body);
+    return visitor.set;
+
+    struct EscapeVisitor {
+        escaping: bool,
+        set: FxHashSet<Local>,
+    }
+
+    impl Visitor<'_> for EscapeVisitor {
+        fn visit_local(&mut self, local: &Local, _: PlaceContext, _: Location) {
+            if self.escaping {
+                self.set.insert(*local);
+            }
+        }
+
+        fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+            if place.projection.is_empty()
```

Comment:
I think we should play safe and reject all locals that have their address taken. Whether only for a part of as a whole. At least until we have better support for handling references in mir opts.